changelog.d/12522.bugfix

```diff
@@ -0,0 +1 @@
+Fix a bug introduced in Synapse 0.99.3 which could cause Synapse to consume large amounts of RAM when back-paginating in a large room.
```

Comment:
Suppose we're in the worst case, where we have many backfill points to consider and none of them/their successor events are visible to us. Wouldn't we end up doing the same amount of work as before this change? (I.e. isn't this a bugfix for the best/average case rather than the worst?)

Reply:
> Wouldn't we end up doing the same amount of work as before this change?

I think we won't try to load all state maps into memory *at the same time* (which we used to do in the depths of `filter_events_for_server`, in `_event_to_memberships`). So yeah, we might end up churning a bunch of CPU which we didn't before, but at least we won't OOM.